### PR TITLE
Ensure container/model is ready to be queried

### LINKED ISSUE #152: Clipper returns default predictions when a new version of a model is deployed but not connected
When a user updates a model version, Clipper is informed that the model version has changed and it immediately starts to query the new version. However, it can take several minutes for the container for that new version to initialize and connect to Clipper. In that intervening period, Clipper attempts to query the latest version, cannot find any containers for that model version, and instead returns the default prediction.

Instead, it may be desirable for Clipper to wait until the new container has finished initializing and connects to Clipper before switching to the new version.

### LINKED ISSUE #425: Signal when rpc.start() is called
Use Docker health checking to signal when rpc.start() is called.

### LINKED ISSUE #436: [Kubernetes] What image pull policy to use?
- So we do not have a image pull policy in all Kuberntetes configuration. This will lead to Kubernetes use the cached image on the node. This is fine for Clipper's management/query frontend as well as redis. **However, model container shouldn't be cached**. 
- There is model version exposed to the user but the most likely scenario is: user use the same name for some model over and over again and probably not paying much attention to the version. 
- Fix is simple: add `imagePullPolicy: Always` somewhere [here](https://github.com/ucbrise/clipper/blob/bd579202b3ef103802c8733d4cf0672669f0841f/clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py#L188-L216)

### PR DESCRIPTION
**Pull request by chester-leung**
Implemented Docker healtcheck and Kubernetes readiness probe to ensure that a container/model is ready to be queried. This change blocks `deploy_model()` until we get explicit notification that the container/model is running. 

Also changes pull policy in query frontend and management frontend such that new Docker images are always pulled.

Fixes #436, Fixes #425, Fixes #152

### CODE REVIEW

**Comment by chester-leung** [comment]
For whatever reason, old changes that have already been merged still show up. New changes related to the issues I'm fixing are from b5e3b0c down.

**Comment by dcrankshaw** [comment]
@chester-leung I cleaned up the commit history to have only the new commits. Reviewing now.

File: bin/build_docker_images.sh [hunk 1]
```diff
@@ -241,29 +241,29 @@ build_images () {
...
+    #create_image pyspark-container PySparkContainerDockerfile $public
+    #create_image tf_cifar_container TensorFlowCifarDockerfile $public
+    #create_image tf-container TensorFlowDockerfile $public
+    #create_image pytorch-container PyTorchContainerDockerfile $public
+    #create_image caffe2-onnx-container Caffe2OnnxDockerfile $public
+    #create_image mxnet-container MXNetContainerDockerfile $public
 
     # Build Metric Monitor image - no dependency
-    create_image frontend-exporter FrontendExporterDockerfile $public
+    #create_image frontend-exporter FrontendExporterDockerfile $public
```
Comment (dcrankshaw) [inline bin/build_docker_images.sh]: Uncomment all the commented out docker image builds

File: clipper_admin/clipper_admin/docker/docker_container_manager.py [hunk 2]
```diff
@@ -4,6 +4,8 @@
 import os
 import random
 import time
+import json
+import subprocess
```
Comment (dcrankshaw) [inline clipper_admin/clipper_admin/docker/docker_container_manager.py]: Delete the subprocess import

File: containers/python/rpc.py [hunk 3]
```diff
@@ -511,10 +514,19 @@ def start(self, model, host, port, model_name, model_version, input_type):
         self.server.model_input_type = model_input_type
         self.server.model = model
 
+        # Create a file named model_is_ready.check to show that model and container
+        # are ready
+        # os.system("touch model_is_ready.check")
```
Comment (dcrankshaw) [inline containers/python/rpc.py]: Delete the commented out line

File: containers/python/rpc.py [hunk 4]
```diff
@@ -511,10 +514,19 @@ def start(self, model, host, port, model_name, model_version, input_type):
...
+        with open("/model_is_ready.check", "w") as f:
+            f.write("READY")
         child_conn, parent_conn = Pipe(duplex=False)
         metrics_proc = Process(target=run_metric, args=(child_conn, ))
         metrics_proc.start()
-        self.server.run(parent_conn)
+        try:
+            self.server.run(parent_conn)
+        except Exception as e:
+            os.system("rm model_is_ready.check")
```
Comment (dcrankshaw) [inline containers/python/rpc.py]: Use a Python method and also the path is wrong. Change to:
```py
os.remove("/model_is_ready.check")
```

File: dockerfiles/RPCDockerfile [hunk 5]
```diff
@@ -19,4 +19,6 @@ COPY monitoring/metrics_config.yaml /container/
 
 ENV CLIPPER_MODEL_PATH=/model
 
+HEALTHCHECK --interval=3s --timeout=3s --retries=1 CMD test -f /model_is_ready.check || exit 1
```
Comment (dcrankshaw) [inline dockerfiles/RPCDockerfile]: What is the `retries` field for?

File: dockerfiles/SumDockerfile [hunk 6]
```diff
@@ -5,4 +5,5 @@ COPY containers/python/sum_container.py /container/
 
 CMD ["python", "/container/sum_container.py"]
 
+HEALTHCHECK --interval=3s --timeout=3s --retries=1 CMD test -f model_is_ready.check
```
Comment (dcrankshaw) [inline dockerfiles/SumDockerfile]: Delete, since this is already in the `RPCDockerFile`

File: dockerfiles/RPCDockerfile (hunk 5 above)
Comment (dcrankshaw) [inline dockerfiles/RPCDockerfile]: @jwirjo:  Note the change to the RPCDockerfile. Make sure you include this in the Py3RPCDockerfile

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
Looks good. Just some small comments.

File: clipper_admin/clipper_admin/docker/docker_container_manager.py (hunk 2 above)
Comment (chester-leung) [reply clipper_admin/clipper_admin/docker/docker_container_manager.py]: Done

File: bin/build_docker_images.sh (hunk 1 above)
Comment (chester-leung) [reply bin/build_docker_images.sh]: done

File: containers/python/rpc.py (hunk 3 above)
Comment (chester-leung) [reply containers/python/rpc.py]: Done

File: containers/python/rpc.py (hunk 4 above)
Comment (chester-leung) [reply containers/python/rpc.py]: Done

File: dockerfiles/RPCDockerfile (hunk 5 above)
Comment (chester-leung) [reply dockerfiles/RPCDockerfile]: "It takes `retries` consecutive failures of the health check for the container to be considered `unhealthy`"

File: dockerfiles/SumDockerfile (hunk 6 above)
Comment (chester-leung) [reply dockerfiles/SumDockerfile]: Done

**Comment by dcrankshaw** [comment]
Jenkins test this please

**Comment by dcrankshaw** [comment]
@chester-leung what's the status of this? Are you waiting on a review?

**Comment by chester-leung** [comment]
@dcrankshaw I'm trying to figure out why the build is failing - the tests are all passing when I run them locally. I encountered a problem locally similar to the current error when I hadn't rebuilt the py-rpc docker image to have the healthcheck. I think somehow the healthcheck isn't being set for the containers when Jenkins runs the tests.

**Comment by chester-leung** [comment]
Jenkins test this please

**Review by dcrankshaw — APPROVED** [review]
lgtm

**Pull request merged**